### PR TITLE
[AKS] Add a new command to enable bootstrap artifact source when creating cluster

### PR DESCRIPTION
**Pull request by bingosummer**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [x] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py [hunk 1]
```diff
@@ -3253,6 +3265,25 @@ def set_up_agentpool_profile_ssh_access(self, mc: ManagedCluster) -> ManagedClus
                 agent_pool_profile.security_profile.ssh_access = ssh_access
         return mc
 
+    def set_up_bootstrap_profile(self, mc: ManagedCluster) -> ManagedCluster:
+        self._ensure_mc(mc)
+
+        bootstrap_artifact_source = self.context.get_bootstrap_artifact_source()
+        bootstrap_container_registry_resource_id = self.context.get_bootstrap_container_registry_resource_id()
+        if bootstrap_artifact_source is not None:
+            if bootstrap_artifact_source == CONST_ARTIFACT_SOURCE_DIRECT and bootstrap_container_registry_resource_id:
```
Comment (fseldow) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:3274]: what about `bootstrap_artifact_source != CONST_ARTIFACT_SOURCE_CACHE`? There might be other artifact source in the future.

Comment (fseldow) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:3274]: what about `bootstrap_artifact_source != CONST_ARTIFACT_SOURCE_CACHE`? There might be other artifact source in the future.

Comment (bingosummer) [reply src/aks-preview/azext_aks_preview/managed_cluster_decorator.py]: fixed.

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py [hunk 2]
```diff
@@ -3268,6 +3279,25 @@ def set_up_agentpool_profile_ssh_access(self, mc: ManagedCluster) -> ManagedClus
...
+        if bootstrap_artifact_source is not None:
+            if bootstrap_artifact_source != CONST_ARTIFACT_SOURCE_CACHE and bootstrap_container_registry_resource_id:
+                raise MutuallyExclusiveArgumentError(
+                    "Cannot specify --bootstrap-container-registry-resource-id when "
+                    "--bootstrap-artifact-source is not Cache."
+                )
+            if mc.bootstrap_profile is None:
+                mc.bootstrap_profile = self.models.ManagedClusterBootstrapProfile()  # pylint: disable=no-member
+            mc.bootstrap_profile.artifact_source = bootstrap_artifact_source
+            if bootstrap_artifact_source == CONST_ARTIFACT_SOURCE_CACHE:
```
Comment (karataliu) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:3296]: >CONST_ARTIFACT_SOURCE_CACHE [](http://example.com/codeflow?start=44&length=27)

no need check? already checked bootstrap_container_registry_resource_id must be None when not cache

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py [hunk 3]
```diff
@@ -2652,6 +2653,16 @@ def get_ssh_access(self) -> Union[str, None]:
         """
         return self.raw_param.get("ssh_access")
 
+    def get_bootstrap_artifact_source(self) -> Union[str, None]:
+        """Obtain the value of bootstrap_artifact_source.
+        """
+        return self.raw_param.get("bootstrap_artifact_source")
+
+    def get_bootstrap_container_registry_resource_id(self, enable_validation: bool = False) -> Union[str, None]:
```
Comment (karataliu) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:2661]: >enable_validation [](http://example.com/codeflow?start=59&length=17)

remove? not used

**Review by karataliu — CHANGES_REQUESTED** [review]
:clock1:

**Review by fseldow — APPROVED** [review]

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py (hunk 3 above)
Comment (bingosummer) [reply src/aks-preview/azext_aks_preview/managed_cluster_decorator.py]: removed

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py [hunk 4]
```diff
@@ -13530,3 +13530,212 @@ def test_aks_check_network(self, resource_group, resource_group_location):
         # delete
         self.cmd(
             'aks delete -g {resource_group} -n {name} --yes --no-wait', checks=[self.is_empty()])
+
+    @live_only()
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:13534]: Seems you've committed the corresponding recording file and CI passed, it's not necessary to add this mark.

If you mean the feature is protected by some feature flag or toggle and cannot be bypassed via custom feature header, could remove the case from default test matrix used by cli runner. `src/aks-preview/azcli_aks_live_test/configs/ext_matrix_default.json`

**Review by FumingZhang — COMMENTED** [review]
Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=91161035&view=results)
- test_aks_artifact_source

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py (hunk 4 above)
Comment (bingosummer) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:13535]: live_only is required due to role assignment is not mocked

**Review by karataliu — APPROVED** [review]

**Comment by yonzhan** [comment]
Please fix conflicting files.

**Comment by bingosummer** [comment]
> Please fix conflicting files.

fixed.

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**